xdsl/pattern_rewriter.py

```diff
@@ -439,6 +457,97 @@ def impl(self: _RewritePatternT, op: Operation, rewriter: PatternRewriter) -> No
     return impl
 
 
+class TypeConversionPattern(RewritePattern):
+    @abstractmethod
+    def convert_type(self, typ: Attribute, /) -> Attribute | None:
+        ...
+
+    def convert_type_rec(self, typ: Attribute) -> Attribute | None:
+        inp = typ
+        if isinstance(typ, ParametrizedAttribute):
+            parameters = list(self.convert_type_rec(p) or p for p in typ.parameters)
+            inp = type(typ).new(parameters)
+        if isa(typ, ArrayAttr[Attribute]):
+            parameters = tuple(self.convert_type_rec(p) or p for p in typ)
+            inp = type(typ).new(parameters)
+        return self.convert_type(inp) or inp
+
+    @final
+    def match_and_rewrite(self, op: Operation, rewriter: PatternRewriter):
+        new_result_types: list[Attribute] = []
+        new_attributes: dict[str, Attribute] = {}
+        changed: bool = False
+        for result in op.results:
+            converted = self.convert_type_rec(result.type)
+            new_result_types.append(converted or result.type)
+            if converted and converted != result.type:
+                changed = True
+        for name, attribute in op.attributes.items():
+            converted = self.convert_type_rec(attribute)
+            new_attributes[name] = converted or attribute
+            if converted and converted != attribute:
+                changed = True
+        for region in op.regions:
+            for block in region.blocks:
+                for arg in block.args:
+                    converted = self.convert_type(arg.type)
+                    if converted and converted != arg.type:
+                        rewriter.modify_block_argument_type(arg, converted)
```

Comment:
missing `changed=True`?

Reply:
No, because here I do the change in-place through the rewriter if needed.
This is al to go through the rewriter API, mostly for the statekeeping it does for recursive application
(This way, the rewrite infrastructure knows when applying this pattern didn't change anything and won't infinitely recurse!)

Reply:
So I record changed for things that would require to replace the op below